\begin{abstract}
RNA sequencing is a popular asset for measuring transcriptomes and it is useful for expression quantification and sequence assembly of genes and transcripts. One of the primary computational challenges in assembly is the vast amount of data that puts a burden on both the runtime of assembly algorithms and memory usage of the data structures that we adopt for it. Not limited to assembly, the data structures that are used for fast mapping and alignment and their indexing scheme are also memory consuming and fail to scale to sequences of the size of the genome or collection of genomes. We have developed new data structures to substantially reduce the memory used by these indices and still achieve similar query speed. These are highly promising data structures for fast assembly, sequence alignment or mapping, and are useful for genomic, metagenomic, and microbiome analysis.

Specifically we designed a new, succinct representation for the color information of colored de Bruijn graphs (in a tool called “rainbowfish”). This representation can be used in de novo assembly and variant detection to keep information about the sample of origin of each k-mer when combining many samples. For example, our data structure allows build a colored de Bruijn graph on a metagenomic data set in just a few gigabytes while the space state-of-the-art data structures take is hundreds of gigabytes, showing an order-of-magnitude improvement. Moreover, the query time for searching a k-mer and fetching the color information in our representation is almost the same as the existing data structures. We’ve also developed a new data structure for indexing the compacted, colored de Bruijn graph (implemented in a tool called “pufferfish”), which can be used in indexing large-scale reference sequences, or large collections of reference sequences. We have developed both a sparse and dense indexing scheme which allows one to trade off index space for query speed (though queries always remain asymptotically optimal). In addition to indexing the k-mers, we store informations such as position and orientation of each k-mer in the reference set. This space-efficient data structure is built on top of the compacted representation of de Bruijn graph and hence will work well as an index for purposes such as alignment and quantification. In designing both of these methods, we make use of techniques from succinct data structure design — building more complex operations atop bit vectors and the rank and select operations. Pufferfish also makes use of minimum perfect hashing to help substantially reduce the space required for compacted de Bruijn graph indexing.
% de Bruijn graphs are nowadays an inseparable part of Next Generation Sequence analyses. With the fast growth in the amount of sequencing reads and variety of different genomes and transcriptomes the community is shifting toward using graphs to represent collections of references and map reads. 
% In this document, we present two succinct representations for two different variations of a de Bruijn graph, namely colored de Bruijn graph and compacted de Bruijn graph. For the former, we designed and developed rainbowfish, a succinct data structure to represent colors for an efficient \dbg representation and also theoretically proved why we call it succinct. This structure is useful in genome variant detection and genotyping. For the later, we developed a tool named pufferfish for indexing a compacted de Bruijn graph in two different schemes of dense and sparse. While being close to linear indexing methodologies regarding memory, pufferfish shows to have a similar \kmer lookup speed as other memory-consuming de Bruijn graph indexing schemes. The balance that pufferfish offers between memory and speed makes it suitable for large-scale indexing such as for collections of RNA-seq data or in metagenomic analysis.
\end{abstract} 