\begin{abstract}
    \normalsize
RNA sequencing is a popular asset for measuring transcriptomes and
is useful for sequence alignment, expression quantification, and sequence assembly of genes and transcripts.
The vast amount of available raw RNA sequences, and its growth rate over the past decade
brings us with an important
computational challenge prior to any higher level analytical tasks
which is a memory-efficient data structure for indexing the underlying set of sequences
that provides a time-efficient query.

This challenge is not limited to only raw RNA sequences.
By the continuing assembly of novel genomes and transcriptomes,
specifically in metagenomics, having an efficient index over the list of assembled sequences is
still an ongoing challenge. To be specific,
there is still a high demand for an index providing a balance between memory and query time.

We have developed new data structures to substantially reduce the memory used by these indices
and still achieve similar query speed.
These are highly promising data structures for sequence search, sequence alignment, or sequence assembly
and are useful for transcriptomic, genomic, metagenomic, and microbiome analysis.

We can divide these data structures into two categories of reference-based and reference-free indices.
We propose three separate indices,
one over the set of reference sequences used for alignment and mapping applications called Pufferfish,
one over a set of raw sequences designed to be accessed for assembly, variant detection and bubble calling
called Rainbowfish,
and a third one also being over a big set of sequences, but used in sequence search applications called Mantis.
All these data structures are based on de Bruijn graph and different variations of it
such as colored de Bruijn graph or compacted de Bruijn graph.

Specifically we designed a new, succinct representation for the color information
of colored de Bruijn graphs in “Rainbowfish”.
This representation can be used in de novo assembly and variant detection
to keep information about the sample of origin of each k-mer when combining many samples.
For example, our data structure allows build a colored de Bruijn graph
on a metagenomic data set in just a few gigabytes
while the space state-of-the-art data structures take is hundreds of gigabytes,
showing an order-of-magnitude improvement.
Moreover, the query time for searching a k-mer and fetching the color information
in our representation is almost the same as the existing data structures.

Using a similar representation as Rainbowfish to store the list of samples each k-mer appears in
, combined with Counting Quotient Filter to find the index associated to each k-mer we've developed Mantis,
a space-frugal index over thousands of samples with a fast query time up to 106 percent faster than
the sequence search tools at the time.
Mantis is also a colored de Bruijn graph representation,
so it supports fast graph traversal and other topological analyses
in addition to large-scale sequence-level searches.

Later, we update the color information representation
by adopting a hierarchical encoding that exploits correlations among color vectors.
We apply this encoding in the context of two different applications;
the implicit cdbg used for a large-scale sequence search index,
Mantis, as well as the encoding of color information used in
population-level variation detection tool, Rainbowfish.
Our results show significant improvements in the overall size and
scalability of representation of the color information.

In Pufferfish, however, we've developed a new data structure for indexing large-scale reference sequences,
or large collections of reference sequences rather than raw sequence samples by building an index
for the compacted, colored de Bruijn graph over the reference sequences.
We have developed both a sparse and dense indexing scheme which allows one
to trade off index space for query speed
(though queries always remain asymptotically optimal).
In addition to indexing the k-mers, we store informations such as
position and orientation of each k-mer in the reference set.
This space-efficient data structure is built on top of
the compacted representation of de Bruijn graph and hence will work well as an index
for purposes such as alignment and quantification.

In designing all the above methods, we make use of techniques
from succinct data structure design — building more complex operations atop bit vectors
and the rank and select operations and also minimum perfect hashing to help substantially reduce the memory
required for compacted de Bruijn graph indexing.

For my thesis, I propose advancing the ideas explained earlier in two specific, but separate direction.
For the reference-base index scheme, I focus on one of the applications of Pufferfish as the alignment tool
used in estimating abundance of known genomes
    in metagenomic analysis. This work is along the tools such as Bracken and Karp and slightly different from
    metagenomic classification methods such as Kraken. In addition to that, I will continue working on
    the scalability limits of Mantis as a large-sequence-search index by investigating the potentials of
    merging two manti without the need to construct the color information in bit vector format and convert it
    to the hierarchical structure and by controlling the construction memory in a limited range
with respect to the size of the input data.


% de Bruijn graphs are nowadays an inseparable part of Next Generation Sequence analyses. With the fast growth in the amount of sequencing reads and variety of different genomes and transcriptomes the community is shifting toward using graphs to represent collections of references and map reads. 
% In this document, we present two succinct representations for two different variations of a de Bruijn graph, namely colored de Bruijn graph and compacted de Bruijn graph. For the former, we designed and developed rainbowfish, a succinct data structure to represent colors for an efficient \dbg representation and also theoretically proved why we call it succinct. This structure is useful in genome variant detection and genotyping. For the later, we developed a tool named pufferfish for indexing a compacted de Bruijn graph in two different schemes of dense and sparse. While being close to linear indexing methodologies regarding memory, pufferfish shows to have a similar \kmer lookup speed as other memory-consuming de Bruijn graph indexing schemes. The balance that pufferfish offers between memory and speed makes it suitable for large-scale indexing such as for collections of RNA-seq data or in metagenomic analysis.
\end{abstract} 