\chapter{Conclusion}

When working at the scale of whole genomes,
the problem of extending indexing strategies to graphs becomes very important.
In addition to that, indexing more than one sample in the same data structure
adds to the complexity of the problem.
In this document, we presented three data structures
for indexing a collection of genomes, transcriptomes, or sample reads
in the form of a colored de Bruijn graph or a compacted colored de Bruijn graph.
As future direction, we are interested in exploring the application
of one of these data structures, Pufferfish, in metagenomic analyses.

%Another area of interest towards improving space efficiency of \cdbgs
%can be exploring other ways to use fewer colors
%to represent all samples instead of compressing the color matrix. O
%ne idea, we have been discussing with Travis Gagie and Christina Bousher for potential collaborations,
%is to reuse the previously assigned colors in disjoint subgraphs and non-adjacent edges. In other words, we cannot reuse the same color for a different sample for \kmers that are adjacent, but we can have colors that are globally reused. In this case, we need to have a proper mapping from a color to its corresponding samples in different disjoint areas of the graph. As a de Bruijn graph is a special type of directed graphs with each node having at most $4$ incoming and $4$ outgoing edges, the chromatic number of such a graph is at most $9$ (maximum degree + 1~\footnote{Brooks' theorem})~\cite{cranston2015brooks}. Therefore, if we can design an efficient color-to-sample mapping, we can reduce the colors in the graph and skip assigning one bit to each color. In this case, we are approaching the problem from a different direction, rather than compressing the color matrix that we explained in~\ref{chap:rainbowfish}.

In the third chapter, we talked about a new indexing tool, \pufferfish.
The main advantage of a data structure like \pufferfish compared to a linear index
is the ability to efficiently map reads to a population of genomes or individual genomes
with annotated variants.
Current tools that are used for alignment and mapping are either suitable for genome or transcriptome,
but not both. \pufferfish fills the gap by allowing fast and accurate mapping
to a collection of genomes and transcriptomes at the same time.
This ability paves the way for accomplishing applications such as novel exon discovery
or RNA-seq quality control.
Because of the distinction between methods that map to the transcriptome
and those that map to the genome,
we lose information that can be derived by putting both of these mappings together.
One immediate outcome of having short reads mapped to both genome and transcriptome
is in RNA-seq quality control.
If we just look at the transcriptome mapping outcome,
we could simply throw all the non-mapped reads out,
ignoring the fact that not being mapped at all is a different observation than being mapped to an intron.
A large fraction of reads mapping to introns is an evidence that the RNA-seq experiment
failed to provide the required quality.
In our ongoing work Puffalign which is an aligner using the \pufferfish index,
we exploit the effect of indexing genome and transcriptome together and aligning reads to
both of them at the same time in the accuracy of the alignment results.
%Having reads that map to both the genome and transcriptome,
%we can account for such an experiment's failure. Another application of dual mapping,
%that is of biological importance, is looking for any intron retention in an RNA-seq experiment.
%High probability of retaining introns in reads from RNA-seq experiments is known to be associated with certain disease phenotypes~\cite{faustino2003pre,tazi2009alternative}. By mapping the reads to just transcriptomes, we can never be aware of intron retention statistics.

We also talked about sequence search indices and compared them with \mantis
in chapter~\ref{chap:mantis}.
\mantis is an indexing data structure for a massive RNA sequencing database enabling sequence
search queries that is more efficient in space and time compared to other tools
(which are based on sequence bloom trees). It uses Counting Quotient filters
to index \kmers and the representation introduced in \rainbowfish
and explained in chapter~\ref{chap:rainbowfish} to represent the samples containing the \kmer.
which together result in a small exact indexes, on a set of sequencing experiments.
We plan to make it dynamic and scalable to the exponentially
growing volume of RNA sequencing data by incorporating the idea of merging two mantises to build a bigger one.
