%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage{amsthm}
\usepackage[utf8]{inputenc} %unicode support
\usepackage{xspace}
\usepackage{bm}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\newtheorem{thm}{Theorem}
\newtheorem{observation}[thm]{\textbf{Observation}}
\newcommand{\system}{pufferfish\xspace}
\newcommand{\dbg}{de Bruijn graph\xspace}
\newcommand{\cdbg}{compacted de Bruijn graph\xspace}
\newcommand{\kmer}{k-mer\xspace}
\newcommand{\kmers}{k-mers\xspace}
\newcommand{\bv}{\ensuremath{\bm{bv}}\xspace}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Research}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{A space and time-efficient compacted de Bruijn graph index}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   corref={aff1},                       % id of corresponding address, if any
   noteref={n1},                        % id's of article notes, if any
   email={jane.e.doe@cambridge.co.uk}   % email address
]{\inits{JE}\fnm{Jane E} \snm{Doe}}
\author[
   addressref={aff1,aff2},
   email={john.RS.Smith@cambridge.co.uk}
]{\inits{JRS}\fnm{John RS} \snm{Smith}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Department of Zoology, Cambridge}, % university, etc
  \street{Waterloo Road},                     %
  %\postcode{}                                % post or zip code
  \city{London},                              % city
  \cny{UK}                                    % country
}
\address[id=aff2]{%
  \orgname{Marine Ecology Department, Institute of Marine Sciences Kiel},
  \street{D\"{u}sternbrooker Weg 20},
  \postcode{24105}
  \city{Kiel},
  \cny{Germany}
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
\note[id=n1]{Equal contributor} % note, connected to author
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract

  We present a novel data structure for representing the compacted \dbg for use
  as a pattern matching index. As the popularity of the \dbg as an index has
  increased over the past few years, so have the number of proposed
  representations of this structure. Existing structures typically fall into two
  categories; those that are hash-based and provide very fast access to the
  underlying k-mer information and those that are space-frugal and provide
  asymptotically efficient but practically slower pattern search.

  Our representation achieves a compromise between these two extremes. By
  building upon a minimum perfect hash function, we provide practically fast
  k-mer lookup, but by carefully organizing our data structure and making use of
  succinct representations where applicable, we greatly reduce the space
  compared to traditional hashing-based implementations. Further, we provide the
  ability to sample positional information, providing the user with the ability
  to trade off space for speed in a fine-grained manner.

  We believe this representation strikes a desirable balance between speed and
  space usage, and it will allow for fast search and read mapping on large
  reference sequences.
\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{sample}
\kwd{article}
\kwd{author}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Introduction}

The \dbg is a widely-adopted stucture for genome and transcriptome assembly
[cite Pevzner, trinity, ...]. However, the compacted variant of the \dbg recenly
gained increasing use as both an indexing data structure---for use in read
alignment [cite deBGA] and pseudoalignment [cite kallisto]---as well as a
structure for analysis of variation (among multiple genomes) [cite TwoPaCo,
  other colored dBG stuff].

As an indexing data structure, the compacted \dbg is particularly attractive for
dealing with repetitive sequences, since exactly repeated sequences of length at
least k are represented only once in the set of contigs. As has been
demonstrated by Liu et al. [cite deBGA], this considerably speeds up alignment
to repeat-heavy genomes (e.g., the human genome) as well as collections of
related genomes. Further, it has also been demonstrated by Bray et al. [cite
  kallisto], that if full alignment information is not needed, a
properly-indexed compacted \dbg also supports very efficient queries to
determine the reference sequences compatible with sequenced reads (i.e.,
pseudoalignment).

However, the speed of existing compacted \dbg indices compes at a considerable
cost in index size and memory usage. Specifically, the need to build a hash
table over the k-mers appearing in the \dbg contigs requires a large amount of
memory, even for genomes of typical size. Typically, these hash functions map
each k-mer (requiring at least 8 bytes) to the contig in which it occurs
(typically 4 or 8 bytes) and the offset where the k-mer appears in this contig
(again, typically 4 or 8 bytes). A number of other data structures are also
required, but this hash table typically dominates the overall index size. For
example, an index of the human genome constructed in such a manner (i.e., by
deBGA or kallisto) requires $70$---$100$GB of RAM. This already exceeds the
memory requirements of moderate servers (e.g., those with 64G of RAM), and these
requirements quickly become untenable with larger genomes or collections of
genomes.




Text and results for this section, as per the individual journal's instructions for authors. %\cite{koon,oreg,khar,zvai,xjon,schn,pond,smith,marg,hunn,advi,koha,mouse}

\section*{Method}
Text for this section \ldots

\subsection*{The dense \system index}

Here we describe the basic (i.e., dense) \system index. The index consists of 6 components, each of which is described below:

\begin{enumerate}

\item The contig array (CSeq) consists of the (2-bit encoded) sequence of all
  contigs of the \cdbg packed together into a single array. Typically, the size
  of this structure is close to (or smaller than) the size of the 2-bit encoded
  refernece sequence, since redundant sequences are represented only once in
  this structure. We note that the contig array contains the sequence of every
  valid \kmer, as well as that of potentially invalid \kmers (those which span
  contig boundaries in the packed array). We denote by $S$ the total length (in
  nucleotides) of the contig array.

\item The boundary vector (\bv) is a sparse bit-vector with a length of $S$
  bits. The bits of this vector are in one-to-one correspondence with the
  nucleotides of the contig array, and the boundary vector contains a one at
  each nucleotide corresponding to the end of a contig in the contig array, and
  a zero everywhere else.

\item The minimum perfect hash function ($h\left(\right)$) maps every
  \emph{valid} \kmer in the contig array (i.e., all \kmers not spanning contig
  boundaries) to a unique number in $\left[0,N\right)$, where $N$ is the number
    of distinct valid \kmers in CSeq.

\item The position vector (\emph{pos}) stores, for each valid \kmer $x$, the
  position where this \kmer occurs in CSeq. Specifically, for \kmer $x$,
  \emph{pos}$\left[h\left(x\right)\right]$ contains the position of $x$ in CSeq
  such that CSeq$\left[h\left(x\right),h\left(x\right) + k\right] = x$.

\item The conitg table stores, for each contig appearing in CSeq, the reference
  sequences (including offset and orientation) where this contig appears in the
  reference. This is similar to a ``posting list'' in traditional inverted
  indices, where all occurencces of the item (in this case, an entire \cdbg
  contig) are listed. The order of the contigs in the contig table is the same
  as their order in CSeq, allowing the information for a contig to be accessed
  via a simple rank operation on \bv.

\item \emph{Optionally}, an equivalence class table that records, for each
  contig, the set of reference sequences where this contig appears.
  Pre-computation and storage of these equivalence classes can speed up fast
  mapping approaches (e.g., pseudoalignment).
\end{enumerate}


\subsection*{The sparse \system index}

The \system index, as described above, is relatively memory-efficient. Yet, the
biggest component, the \emph{pos} vector, can still grow rather large. This is
because it occupies $\lg(S)$ bits for each of the $N$ valid \kmers in CSeq.
However, at the cost of a slight increase in the practical (though not
asymptotic) complexity of lookup, the size of this structure can be reduced
considerably.  To see how, we first make the following observation:

\begin{observation}
  In the \cdbg (and hence, in CSeq), each valid \kmer occurs exactly once (\kmers occuring between contig boundaries are not considered).
\end{observation}

Hence, any valid \kmer in the \cdbg is either a complex \kmer (i.e., it has an
in our out degree greater than 1), is a terminal \kmer (i.e., it appears at the
beginning or end of some input reference sequence), or it has a unique
predecessor and successor in the orientation defined by the contig.

We can exploit this observation in \system to allow \emph{sampling} of the \kmer
positions. That is, rather than storing the position of each \kmer in the contig
array, we store the position only for some subset of \kmers, where the rate of
sampling is defined by a user-defined parameter $s$. For those \kmers that are
not sampled, we store, instead, three pieces of information; the left or right
extension that must be applied to yield the closest \kmer at a sampled position,
whether or not the corresponding \kmer in CSeq is the canonical version and
whether the extension to reach the nearest sampled position should be applied by
moving to the left or the right. This idea of sampling the positions for the
\kmers is similar to the idea of sampling the suffix array positions that is
employed in the FM-index [cite FM-index]. This allows us to trade off query time
for index space, to allow the \system index to scale to large genomes or
collections of genomes.

\subsection*{Sub-heading for section}
Text for this sub-heading \ldots
\subsubsection*{Sub-sub heading for section}
Text for this sub-sub-heading \ldots
\paragraph*{Sub-sub-sub heading for section}
Text for this sub-sub-sub-heading \ldots
In this section we examine the growth rate of the mean of $Z_0$, $Z_1$ and $Z_2$. In
addition, we examine a common modeling assumption and note the
importance of considering the tails of the extinction time $T_x$ in
studies of escape dynamics.
We will first consider the expected resistant population at $vT_x$ for
some $v>0$, (and temporarily assume $\alpha=0$)
%
\[
 E \bigl[Z_1(vT_x) \bigr]= E
\biggl[\mu T_x\int_0^{v\wedge
1}Z_0(uT_x)
\exp \bigl(\lambda_1T_x(v-u) \bigr)\,du \biggr].
\]
%
If we assume that sensitive cells follow a deterministic decay
$Z_0(t)=xe^{\lambda_0 t}$ and approximate their extinction time as
$T_x\approx-\frac{1}{\lambda_0}\log x$, then we can heuristically
estimate the expected value as
%
\begin{eqnarray}\label{eqexpmuts}
E\bigl[Z_1(vT_x)\bigr] &=& \frac{\mu}{r}\log x
\int_0^{v\wedge1}x^{1-u}x^{({\lambda_1}/{r})(v-u)}\,du
\nonumber\\
&=& \frac{\mu}{r}x^{1-{\lambda_1}/{\lambda_0}v}\log x\int_0^{v\wedge
1}x^{-u(1+{\lambda_1}/{r})}\,du
\nonumber\\
&=& \frac{\mu}{\lambda_1-\lambda_0}x^{1+{\lambda_1}/{r}v} \biggl(1-\exp \biggl[-(v\wedge1) \biggl(1+
\frac{\lambda_1}{r}\biggr)\log x \biggr] \biggr).
\end{eqnarray}
%
Thus we observe that this expected value is finite for all $v>0$ (also see \cite{koon,khar,zvai,xjon,marg}).
%\nocite{oreg,schn,pond,smith,marg,hunn,advi,koha,mouse}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
    Text for this section \ldots

\section*{Acknowledgements}
  Text for this section \ldots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc_article}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      A short description of the figure content
      should go here.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      Figure legend text.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}
\begin{table}[h!]
\caption{Sample table title. This is where the description of the table should go.}
      \begin{tabular}{cccc}
        \hline
           & B1  &B2   & B3\\ \hline
        A1 & 0.1 & 0.2 & 0.3\\
        A2 & ... & ..  & .\\
        A3 & ..  & .   & .\\ \hline
      \end{tabular}
\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{backmatter}
\end{document}
